OpenRA.Game/Graphics/WorldRenderer.cs

```diff
@@ -133,7 +133,27 @@ void GenerateRenderables()
 			foreach (var e in World.ScreenMap.RenderableEffectsInBox(Viewport.TopLeft, Viewport.BottomRight))
 				renderablesBuffer.AddRange(e.Render(this));
 
-			renderablesBuffer.Sort((x, y) => RenderableZPositionComparisonKey(x).CompareTo(RenderableZPositionComparisonKey(y)));
+			renderablesBuffer.Sort((x, y) =>
+			{
+				// Sort is an unstable sort, so elements with the same Z position may get sorted differently each frame.
+				// This leads to flickering when rendering several frames as the elements get swapped at random.
+				// To combat this, use a series of arbitrary tie-breaks to determine the "top-most" element.
+				// This ensures a consistent decision for which one is on top, preventing flicker.
+				var compared = RenderableZPositionComparisonKey(x).CompareTo(RenderableZPositionComparisonKey(y));
+				if (compared != 0)
+					return compared;
+
+				compared = x.Pos.X.CompareTo(y.Pos.X);
+				if (compared != 0)
+					return compared;
+
+				compared = x.Pos.Y.CompareTo(y.Pos.Y);
+				if (compared != 0)
+					return compared;
+
+				compared = x.Pos.Z.CompareTo(y.Pos.Z);
+				return compared;
```

Comment:
This can simplify to `return x.Pos.Z.CompareTo(y.Pos.Z)`.

Reply:
Also, wouldn't it make sense to compare z before x or y?

Reply:
Simplified.

I chose this order as X is often a useful first-element that actually differs. Z often is no help. That said, it is arbitrary so I can switch it if that's better in some way?

Reply:
We can leave it like this then.